src/state/slices/portfolioSlice/portfolioSlice.ts

```diff
@@ -53,38 +53,42 @@ export const portfolio = createSlice({
         getMixPanel()?.track(MixPanelEvents.DisconnectWallet)
       }
     },
-    upsertAccountMetadata: (state, { payload }: { payload: AccountMetadataById }) => {
-      moduleLogger.debug('upserting account metadata')
-      state.accountMetadata.byId = {
-        ...state.accountMetadata.byId,
-        ...payload,
-      }
+    upsertAccountMetadata: {
+      reducer: (state, { payload }: { payload: AccountMetadataById }) => {
+        state.accountMetadata.byId = merge(state.accountMetadata.byId, payload)
+        state.accountMetadata.ids = Object.keys(state.accountMetadata.byId)
 
-      state.accountMetadata.ids = Object.keys(state.accountMetadata.byId)
+        if (!state.walletId) return // realistically, at this point, we should have a walletId set
+        const existingWalletAccountIds = state.wallet.byId[state.walletId] ?? []
+        const newWalletAccountIds = Object.keys(payload)
+        // keep an index of what account ids belong to this wallet
+        state.wallet.byId[state.walletId] = Array.from(
+          new Set(existingWalletAccountIds.concat(newWalletAccountIds)),
+        )
+      },
```

Comment:
If we feel like it
https://measurethat.net/Benchmarks/Show/8583/0/lodash-uniq-vs-arrayfromnew-set-vs-spread-new-set